### PR TITLE
Remove SASS from @shopify/polaris

### PR DESCRIPTION
**Pull request by jesstelford**
Part of https://github.com/Shopify/polaris-internal/issues/1345

- [x] I've tophatted [the storybook](https://5d559397bae39100201eedc1-bldfeadhvc.chromatic.com/)
- [x] I've tophatted this [on staging](https://web.docs.shopify.io/docs/guides/staging#testing-your-code-in-staging) `staging-sxpt`

---

## Reviewing notes

The majority of changes in this PR are file renames (`.scss` -> `.css` & their imports), which makes it hard to review the changes :(

To view the config related changes, use this command:

```bash
git diff main..HEAD -- ':!polaris-react/src/*' ':!polaris-react/postcss-mixins/*' ':!*.scss'
```

To see what the diff of the final generated .css is, read on...

### Diff details

As well as Tophatting Storybook & web, another way to confirm nothing has broken / changed is comparing the built `styles.css` file agains [the current version](https://unpkg.com/browse/@shopify/polaris/build/esm/styles.css).

The output is not identical, but we didn't expect it to be (eg; moving to CSS Custom Properties). We developed [a script](https://github.com/Shopify/polaris/blob/remove-sass-compiled-css/go.sh) to normalize the output between the two built `.css` files (one from `main`, the other from this `remove-sass` branch) which minimizes the diff to what's actually changed.

For the full diff with inline comments on what's changed and why, see https://github.com/Shopify/polaris/pull/11718/files 

### Performance

**tl;dr:** Building is now 6 seconds _slower_ with `postcss`.

Performance test ran with [hyperfine](https://github.com/sharkdp/hyperfine):

```
❯ hyperfine --parameter-list branch f9d5f25b9,f6ba2b2a8 --setup='git checkout {branch} && rm -rf node_modules */node_modules\(N\) && . ~/.nvm/nvm.sh && nvm use && yarn --prefer-offline' --prepare='rm -rf polaris-for-vscode/dist polaris-icons/dist polaris-migrator/dist polaris-react/build polaris-react/build-internal polaris-tokens/dist polaris.shopify.com/.next' --runs 5 --warmup 1 '. ~/.nvm/nvm.sh && nvm use && yarn build --filter="@shopify/polaris" --force --no-cache'
```

```
Benchmark 1: (SASS)
  Time (mean ± σ):     27.781 s ±  0.641 s    [User: 40.161 s, System: 5.228 s]
  Range (min … max):   27.087 s … 28.579 s    5 runs

Benchmark 2: (postcss)
  Time (mean ± σ):     33.619 s ±  1.117 s    [User: 46.152 s, System: 6.066 s]
  Range (min … max):   32.103 s … 34.852 s    5 runs

Summary
  SASS ran 1.21 ± 0.05 times faster than postcss
```

### CODE REVIEW

File: .changeset/slimy-kings-train.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@shopify/stylelint-polaris': minor
+---
+
+Can override polaris/coverage rules via secondary option: `rules: { 'polaris/coverage': [true, { /* overrided */ }] }`
```
Comment (chloerice) [inline .changeset/slimy-kings-train.md]:
```suggestion
Added support for configuring and overriding polaris/coverage rules with secondary options, e.g., 
    ```js
    rules: { 
      'polaris/coverage': [true, {
        layout: [{
          'declaration-property-value-disallowed-list': [{position: "fixed"}]
        }]
      }]
    }
    ```
```

File: .stylelintrc.js [hunk 2]
```diff
@@ -17,6 +19,37 @@ module.exports = {
...
       },
     },
+    {
+      files: ['polaris-react/postcss-mixins/*.css'],
+      rules: {
+        'comment-word-disallowed-list': [
+          /.*/,
+          {
+            message:
+              'Comments in polaris-react/postcss-mixins/*.css cause postcss-mixins to error. To disable lint rules, add them to styleslintrc.js overrides.',
```
Comment (chloerice) [inline .stylelintrc.js:29]: 👍🏽

File: stylelint-polaris/plugins/coverage/index.js [hunk 3]
```diff
@@ -1,7 +1,445 @@
...
+  'pc',
+  'pt',
+];
+
+/**
+ * @type {import('./plugins/coverage').PrimaryOptions} The stylelint-polaris/coverage rule expects a 3-dimensional rule config that groups Stylelint rules by coverage categories. It reports problems with dynamic rule names by appending the category to the coverage plugin's rule name
+
+(e.g., Unexpected named color "blue" - Please use a Polaris color token Stylelint(polaris/colors/color-named)")
+*/
+const stylelintPolarisCoverageOptions = {
```
Comment (chloerice) [inline stylelint-polaris/plugins/coverage/index.js]: Can/should the rule's configuration move into its own file instead of into the rule's plugin file?

**Review by chloerice — COMMENTED** [review]
Man oh man, this was a non-trivial load of work to accomplish, well done y'all. Code looks good. In order to feel confident in shipping this, it needs to be on a web staging environment. I'll draft a PR and cross-link to this one so folks can tophat on any store.

**Comment by chloerice** [comment]
> 🫰✨ **Thanks @jesstelford! Your snapshots have been published to npm.**
> 
> Test the snapshots by updating your `package.json` with the newly published versions:
> 
> ```shell
> yarn add @shopify/polaris-migrator@0.0.0-snapshot-20240312074904
> ```
> 
> ```shell
> yarn add @shopify/stylelint-polaris@0.0.0-snapshot-20240312074904
> ```

Does something need to change in the publish action along with rest of these changes? `@shopify/polaris` is not getting published with the other snapshots in spite of the majority of changes being in that package 👀 Or maybe it's just because there's no changeset targeting it 🤔 Pushed up a change log draft for `@shopify/polaris` to hopefully unblock getting this branch onto staging in web.

**Comment by alex-page** [comment]
> Or maybe it's just because there's no changeset targeting it

Yeah you need a changeset otherwise it doesn't think this PR has any changes and just snapshots main

File: stylelint-polaris/plugins/coverage/index.js (hunk 3 above)
Comment (jesstelford) [reply stylelint-polaris/plugins/coverage/index.js]: Yeah, definitely can - I just wholesale copy+pasted it from the stylelint-polaris `index` file 🤷

**Comment by jesstelford** [comment]
Thanks for getting the snapshot release sorted @chloerice!

File: polaris-react/playground/DetailsPage.module.css [hunk 4]
```diff
@@ -1,10 +1,9 @@
-@import '../src/styles/common';
-
 .ContextControl {
   display: flex;
   align-items: center;
   padding: 0 var(--p-space-300);
-  height: $top-bar-height;
+  // stylelint-disable-next-line -- polaris custom global property
+  height: var(--pg-top-bar-height);
```
Comment (alex-page) [inline polaris-react/playground/DetailsPage.module.css:6]: I haven't seen `--pg-` before I couldn't quicky find it in this PR but where are these defined?

**Review by alex-page — APPROVED** [review]

File: polaris-react/playground/DetailsPage.module.css (hunk 4 above)
Comment (jesstelford) [reply polaris-react/playground/DetailsPage.module.css:6]: All the `--pg-` custom properties are [defined in `polaris-react/src/components/AppProvider/global.css`](https://github.com/Shopify/polaris/pull/11677/files#diff-8b82a29970607f740c1b4c6ef27122c4b2218cfd923ea7aaba9db69937c8bc6d) (previously `.scss`) - they're "global" variables (hence the `g`) that used to be global SASS variables and are used across components.

`--pg-` has been added to the linter as a disallowed custom property prefix ([and isn't used anywhere](https://grokt.shopify.io/results?q=%22%5C-%5C-pg%5C-%22) so it's a non-breaking change to the linter).

Comment (alex-page) [reply polaris-react/playground/DetailsPage.module.css:6]: Thanks @jesstelford makes sense to me. I didn't like these as scss variables and I also don't like it now but it makes sense until we refactor some of these global layouts and fix what UI needs these values.

Could we replace any of these with size tokens?

Comment (jesstelford) [reply polaris-react/playground/DetailsPage.module.css:6]: We definitely could replace them with hardcoded values in some cases. But then we'd lose the semantic naming and the relationship between where those values are used.

For example, `--pg-layout-width-inner-spacing-base` & its uses could be replaced with `--p-size-400`, but even though [it's only used in one place](https://github.com/Shopify/polaris/blob/main/polaris-react/src/components/Frame/components/ContextualSaveBar/ContextualSaveBar.module.css#L50-L53), the loss of that semantic information would be a serious regression in my opinion.
